compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1601,12 +1601,19 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
 
             ty::Closure(_, substs) => {
                 // (*) binder moved here
-                let ty = self.infcx.shallow_resolve(substs.as_closure().tupled_upvars_ty());
-                if let ty::Infer(ty::TyVar(_)) = ty.kind() {
-                    // Not yet resolved.
-                    Ambiguous
+
+                if let Ok(tupled_tys) = substs.as_closure().tupled_upvars_ty() {
```

Comment:
Nit: it feels like we may want a helper for this pattern of shallow resolving the `tupled_upvars_ty`, but maybe it doesn't occur that often actually.

Reply:
I don't see too much benefit here, especially since each case Tuple/Infer/Error will need to be handled at each call site. Also, we only do this in 3 spots.